File: crates/compiler/zo-benches/benches/rule-110/rule-110.c
#include <stdio.h>

int main() {
  int width = 31;
  int gens = 15;
  int rule[8] = {0, 1, 1, 1, 0, 1, 1, 0};
  int row[31];
  int next[31];

  for (int i = 0; i < width; i++) {
    row[i] = 0;
  }

  row[width - 1] = 1;

  for (int g = 0; g < gens; g++) {
    for (int p = 0; p < width; p++) {
      if (row[p] == 1) {
        printf("#");
      } else {
        printf(".");
      }
    }
    printf("\n");

    for (int k = 0; k < width; k++) {
      int left = 0;
      if (k > 0) left = row[k - 1];
      int center = row[k];
      int right = 0;
      if (k < width - 1) right = row[k + 1];
      int pattern = left * 4 + center * 2 + right;
      next[k] = rule[pattern];
    }

    for (int i = 0; i < width; i++) {
      row[i] = next[i];
    }
  }

  return 0;
}

File: crates/compiler/zo-benches/benches/rule-110/rule-110.rs
fn main() {
  let width: usize = 31;
  let gens: usize = 15;
  let rule: [i32; 8] = [0, 1, 1, 1, 0, 1, 1, 0];

  let mut row: Vec<i32> = vec![0; width];
  row[width - 1] = 1;

  for _g in 0..gens {
    for p in 0..width {
      if row[p] == 1 {
        print!("#");
      } else {
        print!(".");
      }
    }
    println!();

    let mut next: Vec<i32> = vec![0; width];

    for k in 0..width {
      let left = if k > 0 { row[k - 1] } else { 0 };
      let center = row[k];
      let right = if k < width - 1 { row[k + 1] } else { 0 };
      let pattern = (left * 4 + center * 2 + right) as usize;
      next[k] = rule[pattern];
    }

    row = next;
  }
}
